pkg/util/future/future.go

```diff
@@ -0,0 +1,169 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package future
+
+import "sync/atomic"
+
+// Future represents a value which will become ready in the future.
+// A zero initialized Future is ready to be used, and it represents a promise
+// that the caller will produce value in the future.
+// The caller may arrange for a callback to be invoked when the future
+// becomes ready.
+// Use AwaitableFuture if the caller needs to wait for the future.
+type Future[T any] struct {
+	value atomic.Pointer[futureValue[T]]
+}
+
+// futureValue holds value, along with any callbacks that
+// need to run when value becomes available.
+type futureValue[T any] struct {
+	v         *T
+	whenReady []func(v T)
+}
+
+// ErrorFuture is a handy alias for a future that returns an error.
+type ErrorFuture = Future[error]
+
+// AwaitableFuture is a future that can be waited on.
+type AwaitableFuture[T any] struct {
+	*Future[T]
+	done <-chan struct{}
+}
+
+// MakeCompletedFuture constructs future which returns specified value.
+func MakeCompletedFuture[T any](v T) *Future[T] {
+	f := MakePromise[T]()
+	f.Set(v)
+	return f
+}
+
+// MakePromise constructs future where the caller promises to set the value or
+// an error.
+func MakePromise[T any]() *Future[T] {
+	return &Future[T]{}
+}
+
+// MakeCompletedErrorFuture constructs completed ErrorFuture.
+func MakeCompletedErrorFuture(err error) *ErrorFuture {
+	f := ErrorFuture{}
+	f.Set(err)
+	return &f
+}
+
+// MakeAwaitableFuture wraps underlying future so that the callers can wait
+// for future to become available.
+func MakeAwaitableFuture[T any](f *Future[T]) AwaitableFuture[T] {
+	if ptr := f.value.Load(); ptr != nil && ptr.v != nil {
+		// Future already prepared; no need to wait for it to become ready.
+		return AwaitableFuture[T]{Future: f, done: closedChan}
+	}
+
+	done := make(chan struct{})
+	f.WhenReady(func(v T) {
+		close(done)
+	})
+
+	return AwaitableFuture[T]{Future: f, done: done}
+}
+
+// Get blocks until future is ready and returns it.
+// This method blocks unconditionally.  If the caller needs
+// to respect context cancellation, use Done() method to select
+// on, followed by Get.
+func (f AwaitableFuture[T]) Get() T {
```

Comment:
It might seem like an anti-pattern, but first, let me say that earlier versions did take ctx, which I eventually removed.

The reason for this is that the use of some future instances, such as a *very important*  `future.Future[error]`, would become ambiguous.  Did this function return context.Cancel because caller context was canceled? Or, did it return context.Cancel because the future was prepared with context.Cancel error?

I think this API is also more consistent with context usage in general.  It has method "Done" and it has method "Err", and it doesn't have a method that combines those.  If the caller wants to combine it, the caller is free to to do so via select.
Furthermore, if you are waiting on a future, it's also quite possible that that you'd want to wait for more than just ctx.Done/future.Done -- you might have other cases you might want to select on.  Having Get method take context would preclude that.

In short, I think your suggestion is *very* reasonable, and that's exactly what I had during earlier iterations; but over time, decided against that.  Does that make sense?

Reply:
>Did this function return context.Cancel because caller context was canceled? Or, did it return context.Cancel because the future was prepared with context.Cancel error?

Isn't this true for all context cancellation errors? They can all mask some other error. Consider e.g. a select clause, which will randomly choose either a context cancellation error or some other error. There are no ordering guarantees.

Besides, the return value would be `(T, error)`, so the caller can distinguish between them if needed.

>I think this API is also more consistent with context usage in general. It has method "Done" and it has method "Err", and it doesn't have a method that combines those.

The crucial distinction is that `Err()` doesn't block, it's a polling API. If `Get()` was a polling API I wouldn't have any objections.

As is, it isn't clear to me that the call in `Node.RangeFeed()` will return on context cancellation. To find that out, I now have to go read/trace all the code in `RangeFeedPromise()` to check that it propagates context cancellation properly via the future and that there are no circular dependencies or deadlocks. If it took a context I wouldn't have to go do that, because it would be obvious, and a dangling future would just get garbage collected.

Reply:
I split this API into 2.  Get now doesn't wait -- similar to Err; and Wait takes context and returns T, error.